### PR TITLE
Add basic cross-references to Beam pipeline

### PR DESCRIPTION
**Pull request by schroederc**

### CODE REVIEW

File: kythe/go/serving/pipeline/beam.go [hunk 1]
```diff
@@ -77,6 +79,54 @@ func FromEntries(s beam.Scope, entries beam.PCollection) *KytheBeam {
 	return FromNodes(s, nodes.FromEntries(s, entries))
 }
 
+// CrossReferences returns a Kythe file decorations table derived from the Kythe
+// input graph.  The beam.PCollections haves elements of type
```
Comment (jrtom) [inline kythe/go/serving/pipeline/beam.go]: s/haves/have

Comment (schroederc) [reply kythe/go/serving/pipeline/beam.go]: ✓

File: kythe/go/serving/pipeline/beam.go [hunk 2]
```diff
@@ -89,6 +93,54 @@ func FromEntries(s beam.Scope, entries beam.PCollection) *KytheBeam {
...
+	s := k.s.Scope("CrossReferences")
+	refs := beam.GroupByKey(s, beam.ParDo(s, keyRef, k.References()))
+	// TODO(schroederc): related nodes
+	// TODO(schroederc): callers
+	// TODO(schroederc): MarkedSource
+	// TODO(schroederc): source_node
+	return beam.ParDo2(s, groupCrossRefs, refs)
+}
+
+func groupCrossRefs(key *spb.VName, refStream func(**ppb.Reference) bool, emitSet func(string, *srvpb.PagedCrossReferences), emitPage func(string, *srvpb.PagedCrossReferences_Page)) {
```
Comment (jrtom) [inline kythe/go/serving/pipeline/beam.go]: please add documentation

File: kythe/go/serving/pipeline/beam.go [hunk 3]
```diff
@@ -89,6 +93,54 @@ func FromEntries(s beam.Scope, entries beam.PCollection) *KytheBeam {
...
+		if !ok {
+			g = &srvpb.PagedCrossReferences_Group{Kind: kind}
+			groups[kind] = g
+			set.Group = append(set.Group, g)
+		}
+		g.Anchor = append(g.Anchor, ref.Anchor)
+		set.TotalReferences++
+	}
+
+	sort.Slice(set.Group, func(i, j int) bool { return set.Group[i].Kind < set.Group[j].Kind })
```
Comment (jrtom) [inline kythe/go/serving/pipeline/beam.go]: what is the purpose of sorting these?

File: kythe/go/serving/pipeline/beam.go [hunk 4]
```diff
@@ -89,6 +93,54 @@ func FromEntries(s beam.Scope, entries beam.PCollection) *KytheBeam {
...
+	for refStream(&ref) {
+		kind := refKind(ref)
+		g, ok := groups[kind]
+		if !ok {
+			g = &srvpb.PagedCrossReferences_Group{Kind: kind}
+			groups[kind] = g
+			set.Group = append(set.Group, g)
+		}
+		g.Anchor = append(g.Anchor, ref.Anchor)
+		set.TotalReferences++
```
Comment (jrtom) [inline kythe/go/serving/pipeline/beam.go]: what is total_references actually used for?  that is, why is it a field in the proto, and not just an internal value used by the pipeline to decide when to give up on a particular record?

Comment (schroederc) [reply kythe/go/serving/pipeline/beam.go]: I'm deprecating it; the field isn't useful anymore.

File: kythe/go/serving/pipeline/beam.go (hunk 3 above)
Comment (schroederc) [reply kythe/go/serving/pipeline/beam.go]: To make the output deterministic.

File: kythe/go/serving/pipeline/beam.go (hunk 2 above)
Comment (schroederc) [reply kythe/go/serving/pipeline/beam.go]: ✓

**Pull request merged**